lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -1014,6 +1015,129 @@ class RAIIOpenedArchetypesTracker {
   }
 };
 
+/// Helper routine for propagating concrete type from
+/// class hierarchy analysis.
+bool SILCombiner::propagateConcreteTypeFromCTAInternal(
+    FullApplySite Apply, ProtocolDecl *Protocol, SILValue &Arg,
+    SILValue &NewArg, Optional<ProtocolConformanceRef> &ConformanceRef,
+    ArchetypeType *OpenedArchetype, CanType &ConcreteType) {
+
+  if (VisitedAIForCTA.find(Apply) != VisitedAIForCTA.end())
+    return false;
+  auto &M = Builder.getModule();
+  if (!Protocol)
+    return false;
+  auto *NTD = CTA->getSoleTypeImplementingProtocol(Protocol);
+  if (!NTD)
+    return false;
+
+  ClassDecl *CD = nullptr;
+  /// Check if the class has no subclasses: direct or indirect.
+  if ((CD = dyn_cast<ClassDecl>(NTD)) &&
+          ((CHA->hasKnownDirectSubclasses(CD)) ||
+          (CHA->hasKnownIndirectSubclasses(CD)))) {
+    return false;
+  }
+
+  auto ElementType = NTD->getDeclaredType()->getCanonicalType();
+  if ((isa<ClassDecl>(NTD)) && (!(ElementType->getClassOrBoundGenericClass()))) {
```

Comment:
No need for the extra parentheses. Not clear why this condition occurs, add a comment